pkg/stack/scan.go

```diff
@@ -0,0 +1,74 @@
+// Copyright Nitric Pty Ltd.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stack
+
+import (
+	"go/parser"
+	"go/token"
+	"os"
+	"path/filepath"
+)
+
+var ignore = []string{".", "..", ".git"}
+
+func isMainPackage(file string) (bool, error) {
+	fset := token.NewFileSet()
+	node, err := parser.ParseFile(fset, file, nil, parser.PackageClauseOnly)
+	if err != nil {
+		return false, err
+	}
+	return node.Name.String() == "main", nil
+}
+
+func ScanForHandlers(dir string) ([]string, error) {
+	funcs := []string{}
+	err := filepath.Walk(dir,
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if info.IsDir() || info.Size() == 0 {
+				return nil
+			}
+			for _, i := range ignore {
+				if filepath.Base(info.Name()) == i {
+					return nil
+				}
+			}
+
+			relPath, err := filepath.Rel(dir, path)
+			if err != nil {
+				return err
+			}
+			switch filepath.Ext(info.Name()) {
+			case ".go":
+				isMain, err := isMainPackage(path)
+				if err != nil {
+					return err
+				}
+				if isMain {
+					funcs = append(funcs, relPath)
+				}
+			case ".ts":
```

Comment:
not sure how to figure out if this is a "main" file? Any ideas

Reply:
We might need to use some kind of convention. A few suggestions have been:

Use a directory convention:
e.g. `./functions/**/*.ts`

Or use a file naming convention:
e.g. `./**/*.nitric.ts`

Not sure what everyone prefers though

Reply:
thanks @tjholm , I have updated the PR with a per-runtime config option. This will allow the user to change the default too.